File: pypi/trackfw/commands/serve.py
"""
commands/serve.py — Subcomando `trackfw serve`.
Sobe um servidor HTTP local com dashboard web (stdlib apenas, zero dependências externas).
Espelho Python de internal/commands/serve.go e npm/src/commands/serve.js.
"""

import functools
import ipaddress
import mimetypes
import os
import socket
import sys
import json
from http.server import BaseHTTPRequestHandler, HTTPServer
from urllib.parse import urlparse

STATIC_DIR = os.path.join(os.path.dirname(__file__), "..", "serve", "static")

# Aviso pinado, byte-idêntico entre os 3 runtimes (Go, Node.js, Python) — ver
# docs/cli-parity.md "`trackfw serve` — endereço de escuta, `--host` e
# aviso de exposição". Emitido quando --host resolve para uma interface
# diferente de loopback.
EXPOSURE_WARNING_TEMPLATE = (
    "WARNING: trackfw serve is binding to {host}:{port} — the governance "
    "chain (ADRs, REQs, roadmaps) will be readable without authentication by "
    "any device that can reach it."
)

# Texto de --help pinado, byte-idêntico entre os 3 runtimes.
SERVE_HOST_FLAG_HELP = (
    "Host to bind to (loopback only by default; use 0.0.0.0 to expose on the "
    "network)"
)


def _is_loopback_host(host):
    """Espelha internal/serve/serve.go IsLoopbackHost — 'localhost' ou IP loopback."""
    if host == "localhost":
        return True
    try:
        return ipaddress.ip_address(host).is_loopback
    except ValueError:
        return False


def _exposure_warning(host, port):
    return EXPOSURE_WARNING_TEMPLATE.format(host=host, port=port)


def _display_url(host, port):
    """Espelha internal/serve/serve.go DisplayURL — URL a imprimir/abrir no
    browser. 'localhost' é mantido só para 'localhost' ou IPv4 loopback
    (127.0.0.0/8), para não mudar a saída do caso comum; hosts IPv6 usam
    colchetes; qualquer outro host é impresso como está."""
    if host == "localhost":
        return "http://localhost:{}".format(port)
    try:
        ip = ipaddress.ip_address(host)
    except ValueError:
        return "http://{}:{}".format(host, port)
    if isinstance(ip, ipaddress.IPv6Address):
        return "http://[{}]:{}".format(host, port)
    if ip.is_loopback:
        return "http://localhost:{}".format(port)
    return "http://{}:{}".format(host, port)


def _server_class_for_host(host):
    """Retorna uma subclasse de HTTPServer com address_family AF_INET6 quando
    host é um literal IPv6, e AF_INET (o default da stdlib) caso contrário —
    espelha net.Listen("tcp", net.JoinHostPort(host, port)) do Go, que resolve
    a família do socket a partir do endereço."""
    family = socket.AF_INET
    try:
        if ipaddress.ip_address(host).version == 6:
            family = socket.AF_INET6
    except ValueError:
        pass  # hostnames (ex.: "localhost") ficam em AF_INET, igual ao default da stdlib

    class _Server(HTTPServer):
        address_family = family

    return _Server

# Mapeamento explícito de extensões para garantir Content-Type correto
MIME_TYPES = {
    ".html": "text/html; charset=utf-8",
    ".js": "application/javascript; charset=utf-8",
    ".css": "text/css; charset=utf-8",
    ".json": "application/json",
    ".png": "image/png",
    ".ico": "image/x-icon",
    ".svg": "image/svg+xml",
}


class TrackfwHandler(BaseHTTPRequestHandler):
    """Handler HTTP que serve o dashboard e as APIs REST."""

    def __init__(self, cfg, *args, **kwargs):
        self.cfg = cfg
        super().__init__(*args, **kwargs)

    def do_GET(self):
        parsed = urlparse(self.path)
        path = parsed.path

        if path == "/" or path == "":
            self._serve_static_file("index.html")
        elif path.startswith("/static/"):
            # Remove o prefixo /static/ para obter o nome do arquivo
            filename = path[len("/static/"):]
            # Impedir path traversal em arquivos estáticos
            filename = os.path.basename(filename)
            self._serve_static_file(filename)
        elif path == "/api/board":
            from trackfw.serve.api_board import get_board
            self._json(get_board(self.cfg))
        elif path == "/api/chain":
            from trackfw.serve.api_chain import get_chain
            self._json(get_chain(self.cfg))
        elif path == "/api/metrics":
            from trackfw.serve.api_metrics import get_metrics
            self._json(get_metrics(self.cfg))
        elif path == "/api/file":
            from trackfw.serve.api_file import get_file
            get_file(self.cfg, parsed, self)
        elif path == "/api/attention":
            from trackfw.serve.api_attention import get_attention
            self._json(get_attention(self.cfg))
        else:
            self.send_error(404, "Not found")

    def _json(self, data):
        """Envia resposta JSON."""
        body = json.dumps(data).encode("utf-8")
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(body)

    def _serve_static_file(self, filename):
        """Serve um arquivo do diretório static/."""
        # Normalizar path para evitar traversal
        safe_name = os.path.basename(filename)
        file_path = os.path.join(STATIC_DIR, safe_name)
        file_path = os.path.realpath(file_path)
        static_real = os.path.realpath(STATIC_DIR)

        # Segurança: garantir que o arquivo está dentro de STATIC_DIR
        if not file_path.startswith(static_real + os.sep) and file_path != static_real:
            self.send_error(403, "Access denied")
            return

        if not os.path.isfile(file_path):
            self.send_error(404, f"File not found: {safe_name}")
            return

        ext = os.path.splitext(safe_name)[1].lower()
        content_type = MIME_TYPES.get(ext)
        if not content_type:
            content_type, _ = mimetypes.guess_type(safe_name)
            if not content_type:
                content_type = "application/octet-stream"

        try:
            with open(file_path, "rb") as f:
                body = f.read()
        except OSError as e:
            self.send_error(500, f"Cannot read file: {e}")
            return

        self.send_response(200)
        self.send_header("Content-Type", content_type)
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, format, *args):
        """Silencia os logs padrão do BaseHTTPRequestHandler."""
        pass


def _open_browser(url):
    """Abre o navegador padrão na URL indicada."""
    import subprocess
    import platform

    system = platform.system()
    try:
        if system == "Darwin":
            subprocess.Popen(["open", url])
        elif system == "Windows":
            subprocess.Popen(["start", url], shell=True)
        else:
            subprocess.Popen(["xdg-open", url])
    except OSError:
        pass  # falha silenciosa se não conseguir abrir


def cmd_serve(args):
    """Inicializa e roda o servidor HTTP local."""
    from trackfw import config as _config

    port = getattr(args, "port", 8080)
    host = getattr(args, "host", "127.0.0.1")
    no_open = getattr(args, "no_open", False)

    cfg = _config.load()

    url = _display_url(host, port)

    if not _is_loopback_host(host):
        print(_exposure_warning(host, port), file=sys.stderr)

    # Criar handler com cfg injetado via functools.partial
    handler_class = functools.partial(TrackfwHandler, cfg)
    server_cls = _server_class_for_host(host)

    try:
        server = server_cls((host, port), handler_class)
    except OSError as e:
        print(f"trackfw serve: cannot bind to port {port}: {e}", file=sys.stderr)
        sys.exit(1)

    print(f"trackfw dashboard: {url}")
    print("Press Ctrl+C to stop.")

    if not no_open:
        _open_browser(url)

    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nStopped.")
        server.server_close()
        sys.exit(0)


def register(subparsers):
    """Adiciona subcomando `serve` ao parser principal."""
    parser = subparsers.add_parser(
        "serve",
        help="Start local web dashboard",
    )
    parser.add_argument(
        "--port",
        type=int,
        default=8080,
        metavar="PORT",
        help="Port to listen on (default: 8080)",
    )
    parser.add_argument(
        "--no-open",
        action="store_true",
        dest="no_open",
        help="Do not open browser automatically",
    )
    parser.add_argument(
        "--host",
        type=str,
        default="127.0.0.1",
        metavar="HOST",
        help=SERVE_HOST_FLAG_HELP,
    )
    parser.set_defaults(func=cmd_serve)

File: npm/src/commands/serve.js
'use strict'

const http = require('http')
const path = require('path')
const fs = require('fs')
const { Command } = require('commander')
const config = require('../config')
const { handleBoard } = require('../serve/api_board')
const { handleChain } = require('../serve/api_chain')
const { handleMetrics } = require('../serve/api_metrics')
const { handleFile } = require('../serve/api_file')
const { handleAttention } = require('../serve/api_attention')

const STATIC_DIR = path.join(__dirname, '..', 'serve', 'static')

// REAL_STATIC_DIR é a versão canônica (symlinks resolvidos) de STATIC_DIR.
// Calculado uma vez no carregamento do módulo para canonicalizar ambos os lados
// na verificação de segurança do serveStatic (M-03: prevenção de escape via symlink).
// Se STATIC_DIR não existir no momento do carregamento, fica null e serveStatic
// retorna 404 para qualquer requisição de asset estático.
let REAL_STATIC_DIR
try {
  REAL_STATIC_DIR = fs.realpathSync.native(STATIC_DIR)
} catch (_) {
  REAL_STATIC_DIR = null
}

// Aviso pinado, byte-idêntico entre os 3 runtimes (Go, Node.js, Python) — ver
// docs/cli-parity.md "`trackfw serve` — endereço de escuta, `--host` e
// aviso de exposição". Emitido quando --host resolve para uma interface
// diferente de loopback.
function exposureWarning(host, port) {
  return `WARNING: trackfw serve is binding to ${host}:${port} — the governance chain (ADRs, REQs, roadmaps) will be readable without authentication by any device that can reach it.`
}

// Texto de --help pinado, byte-idêntico entre os 3 runtimes.
const SERVE_HOST_FLAG_HELP = 'Host to bind to (loopback only by default; use 0.0.0.0 to expose on the network)'

// Espelha internal/serve/serve.go IsLoopbackHost — 'localhost' ou IP loopback
// (IPv4 127.0.0.0/8 inteiro, ou IPv6 ::1), igual a net.IP.IsLoopback() do Go
// e ipaddress.ip_address(...).is_loopback do Python.
function isLoopbackHost(host) {
  const net = require('net')
  if (host === 'localhost') return true
  if (net.isIPv4(host)) {
    return host.split('.')[0] === '127'
  }
  if (net.isIPv6(host)) {
    return host === '::1' || host === '0:0:0:0:0:0:0:1'
  }
  return false
}

// Espelha internal/serve/serve.go DisplayURL — URL a imprimir e a abrir no
// browser. 'localhost' é mantido só para 'localhost' ou IPv4 loopback
// (127.0.0.0/8), para não mudar a saída do caso comum; hosts IPv6 usam
// colchetes; qualquer outro host é impresso como está.
function displayUrl(host, port) {
  const net = require('net')
  if (host === 'localhost') return `http://localhost:${port}`
  if (net.isIPv4(host)) {
    return host.split('.')[0] === '127' ? `http://localhost:${port}` : `http://${host}:${port}`
  }
  if (net.isIPv6(host)) {
    return `http://[${host}]:${port}`
  }
  return `http://${host}:${port}`
}

// isValidHost reports whether host is acceptable as a --host argument.
// Accepts: 'localhost', valid IPv4, valid IPv6 literal (without zone ID),
// RFC-1123 hostname. Rejects anything else — in particular strings with shell
// metacharacters, or IPv6 scoped addresses (zone ID after '%').
//
// Zone IDs are rejected even when syntactically clean (e.g. 'fe80::1%eth0')
// because: (1) Go's net.ParseIP rejects them — parity requires all 3 CLIs to
// agree; (2) Python's ipaddress.ip_address() accepts any zone ID content,
// including cmd.exe metacharacters that list2cmdline does not quote; blocking
// '%' aligns both runtimes and closes the entire attack class rather than
// enumerating individual metacharacters.
//
// Espelha internal/serve/serve.go IsValidHost e _is_valid_host do Python.
function isValidHost(host) {
  if (host === 'localhost') return true
  // Reject IPv6 scoped addresses (zone ID): '%' in host means a zone
  // identifier that Python's ipaddress accepts with any content (including
  // cmd.exe metacharacters); Go rejects all scoped addresses; rejecting here
  // makes all three runtimes agree on the same contract.
  if (host.includes('%')) return false
  const net = require('net')
  if (net.isIPv4(host) || net.isIPv6(host)) return true
  // RFC 1123 hostname: labels separated by dots, each [a-zA-Z0-9] or hyphens,
  // starting and ending with alphanum. Max label length 63, total max 253.
  if (host.length > 253) return false
  return /^[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(\.[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/.test(host)
}

// browserArgv returns [cmd, args] for opening url in the default browser
// without shell interpolation — argv, never a shell string. This is the only
// safe form: exec(`open "${url}"`) interpolates url into a shell command and
// allows injection when url contains shell metacharacters.
//
// Darwin / Linux: 'open' / 'xdg-open' are invoked directly; no shell involved.
// Windows: spawn('cmd', ['/c', 'start', '', url]) passes url as a distinct
// argv element to CreateProcess — safer than exec(), but cmd.exe still
// re-parses its own metacharacters (& | ^ >) in the url portion. Defense for
// that residual is AC4 (isValidHost) rejecting metacharacter-bearing hosts
// before they reach this function.
//
// Exported for testing.
function browserArgv(platform, url) {
  if (platform === 'darwin') return ['open', [url]]
  if (platform === 'win32') return ['cmd', ['/c', 'start', '', url]]
  return ['xdg-open', [url]]
}

// openBrowser opens url in the default browser using spawn (argv, no shell).
// Exported for testing with a PATH shim.
//
// Preserves the original UX: warn when the opener exits non-zero (e.g. when
// xdg-open finds no handler) as well as when it cannot be launched (ENOENT).
// The old exec() warned via the callback err; spawn uses 'close' + 'error'.
function openBrowser(platform, url) {
  const { spawn } = require('child_process')
  const [cmd, args] = browserArgv(platform, url)
  const proc = spawn(cmd, args)
  proc.on('error', (err) => {
    console.warn(`Não foi possível abrir o browser: ${err.message}`)
  })
  proc.on('close', (code) => {
    if (code !== 0 && code !== null) {
      console.warn(`Não foi possível abrir o browser: processo encerrou com código ${code}`)
    }
  })
}

// Mapa de extensão → Content-Type
const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.js':   'application/javascript; charset=utf-8',
  '.css':  'text/css; charset=utf-8',
  '.json': 'application/json',
  '.svg':  'image/svg+xml',
  '.ico':  'image/x-icon',
}

/**
 * serveStatic serve arquivos do STATIC_DIR.
 *
 * Modelo de segurança — contenção em dois estágios (M-03):
 *
 * 1. Contenção léxica: path.resolve() normaliza ".." e o resultado deve começar
 *    com STATIC_DIR.  Caminhos fora recebem 403 antes de tocar o disco.
 *
 * 2. Contenção física: fs.realpathSync.native() é chamado no arquivo pedido e
 *    comparado com REAL_STATIC_DIR (pré-computado no carregamento do módulo).
 *    Um symlink dentro de STATIC_DIR cujo destino físico aponta para fora → 403.
 *    Qualquer falha de canonicalização → 404.
 *
 * @param {string} urlPath - pathname da URL (ex: '/static/app.js')
 * @param {http.ServerResponse} res
 */
function serveStatic(urlPath, res) {
  // Remove o prefixo '/static'
  const relative = urlPath.replace(/^\/static/, '') || '/index.html'
  const resolved = path.resolve(path.join(STATIC_DIR, relative))

  // ── Estágio 1: Contenção léxica ────────────────────────────────────────────
  const lexicalStaticDir = path.resolve(STATIC_DIR)
  if (!resolved.startsWith(lexicalStaticDir + path.sep) && resolved !== lexicalStaticDir) {
    res.writeHead(403, { 'Content-Type': 'text/plain' })
    res.end('Forbidden')
    return
  }

  // ── Estágio 2: Contenção física ────────────────────────────────────────────
  if (!REAL_STATIC_DIR) {
    // STATIC_DIR não existe — nenhum arquivo pode ser servido.
    res.writeHead(404, { 'Content-Type': 'text/plain' })
    res.end('Not Found')
    return
  }

  let realResolved
  try {
    realResolved = fs.realpathSync.native(resolved)
  } catch (_) {
    res.writeHead(404, { 'Content-Type': 'text/plain' })
    res.end('Not Found')
    return
  }

  if (!realResolved.startsWith(REAL_STATIC_DIR + path.sep) && realResolved !== REAL_STATIC_DIR) {
    res.writeHead(403, { 'Content-Type': 'text/plain' })
    res.end('Forbidden')
    return
  }

  const ext = path.extname(realResolved).toLowerCase()
  const contentType = MIME[ext] || 'application/octet-stream'

  let content
  try {
    content = fs.readFileSync(realResolved)
  } catch (_) {
    res.writeHead(500, { 'Content-Type': 'text/plain' })
    res.end('Internal Server Error')
    return
  }

  res.writeHead(200, { 'Content-Type': contentType })
  res.end(content)
}

/**
 * createServer cria o servidor HTTP do trackfw serve.
 * @param {object} cfg - configuração do trackfw (resultado de config.load())
 * @param {number} port
 * @returns {http.Server}
 */
function createServer(cfg, port) {
  return http.createServer((req, res) => {
    // CORS permissivo para desenvolvimento local
    res.setHeader('Access-Control-Allow-Origin', '*')

    let urlObj
    try {
      urlObj = new URL(req.url, `http://localhost:${port}`)
    } catch (_) {
      res.writeHead(400)
      res.end('Bad Request')
      return
    }

    const pathname = urlObj.pathname

    if (pathname === '/' || pathname === '/index.html') {
      const indexPath = path.join(STATIC_DIR, 'index.html')
      if (!fs.existsSync(indexPath)) {
        res.writeHead(404, { 'Content-Type': 'text/plain' })
        res.end('index.html not found')
        return
      }
      res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' })
      res.end(fs.readFileSync(indexPath))
      return
    }

    if (pathname.startsWith('/static/')) {
      serveStatic(pathname, res)
      return
    }

    if (pathname === '/api/board') {
      handleBoard(cfg, req, res)
      return
    }

    if (pathname === '/api/chain') {
      handleChain(cfg, req, res)
      return
    }

    if (pathname === '/api/metrics') {
      handleMetrics(cfg, req, res)
      return
    }

    if (pathname === '/api/file') {
      handleFile(cfg, req, res)
      return
    }

    if (pathname === '/api/attention') {
      handleAttention(cfg, req, res)
      return
    }

    res.writeHead(404, { 'Content-Type': 'text/plain' })
    res.end('Not found')
  })
}

/**
 * createServeCommand retorna o comando commander 'serve'.
 * @returns {Command}
 */
function createServeCommand() {
  const cmd = new Command('serve')
  cmd
    .description('Inicia o servidor HTTP do trackfw dashboard (kanban + chain + metrics)')
    .option('--port <port>', 'Porta do servidor', '8080')
    .option('--host <host>', SERVE_HOST_FLAG_HELP, '127.0.0.1')
    .option('--no-open', 'Não abrir o browser automaticamente')
    .action((opts) => {
      const cfg = config.load()
      const port = parseInt(opts.port, 10) || 8080
      const host = opts.host

      // AC4 — validate --host before bind. Rejects metacharacter-bearing
      // strings that would reach the browser-open path or the socket.
      if (!isValidHost(host)) {
        console.error(`trackfw serve: invalid --host value: ${host}`)
        console.error('--host must be "localhost", a valid IPv4/IPv6 address, or an RFC-1123 hostname.')
        process.exit(1)
      }

      const server = createServer(cfg, port)

      if (!isLoopbackHost(host)) {
        console.error(exposureWarning(host, port))
      }

      server.listen(port, host, () => {
        const url = displayUrl(host, port)
        console.log(`trackfw serve: ${url}`)

        if (opts.open !== false) {
          // AC1 — use argv (spawn), never shell string interpolation (exec).
          openBrowser(process.platform, url)
        }
      })

      server.on('error', (err) => {
        if (err.code === 'EADDRINUSE') {
          console.error(`Porta ${port} já está em uso. Use --port para especificar outra.`)
        } else {
          console.error(`Erro no servidor: ${err.message}`)
        }
        process.exit(1)
      })

      // Graceful shutdown
      process.on('SIGINT', () => {
        server.close(() => {
          console.log('\ntrackfw serve: encerrado.')
          process.exit(0)
        })
      })
    })

  return cmd
}

module.exports = { createServeCommand, createServer, isLoopbackHost, displayUrl, isValidHost, browserArgv, openBrowser, serveStatic }
